sycl/include/sycl/ext/oneapi/prefetch.hpp

```diff
@@ -0,0 +1,277 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <CL/__spirv/spirv_ops.hpp>
+#include <sycl/ext/oneapi/properties/properties.hpp>
+
+namespace sycl {
+inline namespace _V1 {
+namespace ext::oneapi::experimental {
+
+enum class cache_level { L1 = 0, L2 = 1, L3 = 2, L4 = 3 };
+
+struct nontemporal;
+
+struct prefetch_hint_key {
+  template <cache_level Level, typename Hint>
+  using value_t =
+      property_value<prefetch_hint_key,
+                     std::integral_constant<cache_level, Level>, Hint>;
+};
+
+template <cache_level Level, typename Hint>
+inline constexpr prefetch_hint_key::value_t<Level, Hint> prefetch_hint;
+
+inline constexpr prefetch_hint_key::value_t<cache_level::L1, void>
+    prefetch_hint_L1;
+inline constexpr prefetch_hint_key::value_t<cache_level::L2, void>
+    prefetch_hint_L2;
+inline constexpr prefetch_hint_key::value_t<cache_level::L3, void>
+    prefetch_hint_L3;
+inline constexpr prefetch_hint_key::value_t<cache_level::L4, void>
+    prefetch_hint_L4;
+
+inline constexpr prefetch_hint_key::value_t<cache_level::L1, nontemporal>
+    prefetch_hint_L1_nt;
+inline constexpr prefetch_hint_key::value_t<cache_level::L2, nontemporal>
+    prefetch_hint_L2_nt;
+inline constexpr prefetch_hint_key::value_t<cache_level::L3, nontemporal>
+    prefetch_hint_L3_nt;
+inline constexpr prefetch_hint_key::value_t<cache_level::L4, nontemporal>
+    prefetch_hint_L4_nt;
+
+namespace detail {
+template <> struct IsCompileTimeProperty<prefetch_hint_key> : std::true_type {};
+
+template <cache_level Level, typename Hint>
+struct PropertyMetaInfo<prefetch_hint_key::value_t<Level, Hint>> {
+  static constexpr const char *name = std::is_same_v<Hint, nontemporal>
+                                          ? "sycl-prefetch-hint-nt"
+                                          : "sycl-prefetch-hint";
+  static constexpr int value = static_cast<int>(Level);
+};
+
+template <typename T, T Parameter, bool AdditionalCondition = true>
+struct prefetch_helper {};
+
+template <access::address_space AddressSpace, bool AdditionalCondition>
+struct prefetch_helper<access::address_space, AddressSpace,
+                       AdditionalCondition> {
+  using type =
+      std::enable_if_t<(AddressSpace == access::address_space::global_space ||
+                        AddressSpace == access::address_space::generic_space) &&
+                           AdditionalCondition,
+                       void>;
+};
+
+template <access_mode AccessMode, bool AdditionalCondition>
+struct prefetch_helper<access_mode, AccessMode, AdditionalCondition> {
+  using type = std::enable_if_t<(AccessMode == access_mode::read ||
+                                 AccessMode == access_mode::write) &&
+                                    AdditionalCondition,
+                                void>;
+};
```

Comment:
I didn't notice this part when I made my suggestion. I wouldn't want to use the same name (`prefetch_helper`) for both, I think we should two helpers with more explicit names.

Also, you can do

```
template <access::address_space AddressSpace, bool AdditionalCondition>
using prefetch_check_as_t = std::enable_if_t<...>;
```
Why don't we allow read-write mode?

Reply:
[Change helper](https://github.com/intel/llvm/pull/11458/commits/9e81d89e11539826a51efffa0bce91811cda2e47)